### PR TITLE
Cfds 137 cap confirmation page

### PR DESCRIPTION
**Pull request by NitishAroraKPMG**
## Description

Cap confirmation page
New page added after the define Cap and will show the user to change any values

## Testing instructions

Create a capped ticket and edit the values in Cap confirmation page.

### CODE REVIEW

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    const capValidity =
+        capValidityAttribute && isCapExpiry(capValidityAttribute)
+            ? `${capValidityAttribute.productValidity}${
+                  capValidityAttribute.productEndTime ? ` - ${capValidityAttribute.productEndTime}` : ''
+              }`
+            : '';
+
+    const capStartAttribute = getSessionAttribute(ctx.req, CAP_START_ATTRIBUTE);
+
+    if (
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:176]: This check needs to happen higher up in the function. We should get all the attributes we know we need, and then do this check to make sure they're all present.

File: repos/fdbt-site/tests/pages/api/capConfirmation.test.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import capConfirmation from '../../../src/pages/api/capConfirmation';
+import { getMockRequestAndResponse } from '../../testData/mockData';
+
+describe('capConfirmation', () => {
+    const writeHeadMock = jest.fn();
+
+    beforeEach(() => {
+        jest.resetAllMocks();
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/pages/api/capConfirmation.test.ts:8]: I dont think we'll need this beforeEach

File: repos/fdbt-site/tests/pages/api/typeOfCap.test.ts [hunk 3]
```diff
@@ -58,9 +58,26 @@ describe('ticketRepresentation', () => {
             Location: '/selectCappedProductGroup',
         });
 
-        expect(updateSessionAttributeSpy).toBeCalledTimes(0);
+        expect(updateSessionAttributeSpy).toBeCalledTimes(1);
     });
-    it('should return 302 redirect to /  defineCapPricingPerDistance when type of cap is byDistance', () => {
+
+    it('should return 302 redirect to /multiTapsPricing when type of cap is byDistance', () => {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/pages/api/typeOfCap.test.ts:64]:
```suggestion
    it('should return 302 redirect to /multiTapsPricing when type of cap is byTaps', () => {
```

File: repos/fdbt-site/tests/pages/api/typeOfCap.test.ts [hunk 4]
```diff
@@ -58,9 +58,26 @@ describe('ticketRepresentation', () => {
...
+            },
+            mockWriteHeadFn: writeHeadMock,
+        });
+        typeOfCap(req, res);
+
+        expect(writeHeadMock).toBeCalledWith(302, {
+            Location: '/multiTapsPricing',
+        });
+
+        expect(updateSessionAttributeSpy).toBeCalledTimes(1);
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/pages/api/typeOfCap.test.ts:77]: We shouldnt be checking the number of times its being called, we should be checking the content.

File: repos/fdbt-site/tests/pages/api/typeOfCap.test.ts [hunk 5]
```diff
@@ -73,6 +90,6 @@ describe('ticketRepresentation', () => {
             Location: '/defineCapPricingPerDistance',
         });
 
-        expect(updateSessionAttributeSpy).toBeCalledTimes(0);
+        expect(updateSessionAttributeSpy).toBeCalledTimes(1);
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/tests/pages/api/typeOfCap.test.ts:93]: Same here.

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        name: 'Cap Expiry',
+        content: sentenceCaseString(capValidity),
+        href: '/selectCapValidity',
+    });
+
+    confirmationElements.push({
+        name: 'Cap Starts',
+        content:
+            capStartInfo.type === 'fixedWeekdays' && capStartInfo.startDay
+                ? `Fixed Days - ${sentenceCaseString(capStartInfo.startDay)}`
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:86]: Fixed days
Rolling days

We use sentence case where we can

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        confirmationElements.push({
+            name: 'Services',
+            content: services.join(', '),
+            href: '/serviceList',
+        });
+    }
+
+    if (tapsPricingContents.length > 0) {
+        confirmationElements.push({
+            name: 'Prices by Taps',
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:101]: Prices by taps

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    capStartInfo,
+    services,
+    tapsPricingContents,
+    capDistancePricingContents,
+    csrfToken,
+}: CapConfirmationProps): ReactElement => (
+    <TwoThirdsLayout title={title} description={description} errors={[]}>
+        <CsrfForm action="/api/capConfirmation" method="post" csrfToken={csrfToken}>
+            <>
+                <h1 className="govuk-heading-l">Check your answers before sending your fares information</h1>
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:132]:
```suggestion
                <h1 className="govuk-heading-l">Check your answers before submitting your cap(s) information</h1>
```

Could you please also change the h1 in faresConfirmation.tsx, from

 <h1 className="govuk-heading-l">Check your answers before sending your fares information</h1>

to

 <h1 className="govuk-heading-l">Check your answers before submitting your fares information</h1>

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        {
+            name: 'Cap type',
+            content: `Pricing by ${typeOfCap.substring(2)} `,
+            href: 'typeOfCap',
+        },
+    ];
+
+    if (productGroupName) {
+        confirmationElements.push({
+            name: 'Product Group Name',
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:58]: Product group name

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+    capValidity: string,
+    capStartInfo: CapStartInfo,
+    services: string[],
+    tapsPricingContents: string[],
+    capDistancePricingContents: string[],
+): ConfirmationElement[] => {
+    const confirmationElements: ConfirmationElement[] = [
+        {
+            name: 'Cap type',
+            content: `Pricing by ${typeOfCap.substring(2)} `,
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:51]: lowerFirst() on the ${typeOfCap.substring(2)}

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+        confirmationElements.push({
+            name: cap.name,
+            content: [`Price - £${cap.price}`, `Duration - ${durationText}`],
+            href: '/createCaps',
+        });
+    });
+
+    confirmationElements.push({
+        name: 'Cap Expiry',
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:77]: Cap expiry

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    const serviceListAttribute = getSessionAttribute(ctx.req, SERVICE_LIST_ATTRIBUTE);
+
+    const multiTapsPricingAttribute = getSessionAttribute(ctx.req, MULTI_TAPS_PRICING_ATTRIBUTE);
+
+    const capDistancePricingAttribute = getSessionAttribute(ctx.req, CAP_PRICING_PER_DISTANCE_ATTRIBUTE);
+
+    const tapsPricingContents: string[] = [];
+    if (multiTapsPricingAttribute && !isWithErrors(multiTapsPricingAttribute)) {
+        Object.entries(multiTapsPricingAttribute.tapDetails).forEach((tap) => {
+            tapsPricingContents.push(`Tap Number: ${Number(tap[0]) + 1}, Price: ${tap[1]}`);
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:195]: Tap number

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        });
+    }
+
+    const capDistancePricingContents: string[] = [];
+    if (capDistancePricingAttribute && !isWithErrors(capDistancePricingAttribute)) {
+        capDistancePricingContents.push(
+            `Max Price: ${capDistancePricingAttribute.maximumPrice}, Min Price: ${capDistancePricingAttribute.minimumPrice}`,
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:202]: Max price Min price

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        );
+        capDistancePricingAttribute.capPricing.forEach((capDistance) => {
+            capDistancePricingContents.push(
+                `Distance: ${capDistance.distanceFrom} - ${capDistance.distanceTo}, Price(per km): ${capDistance.pricePerKm} `,
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:206]: Price per km

File: repos/fdbt-site/src/pages/api/capConfirmation.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import { NextApiResponse } from 'next';
+import { redirectToError, redirectTo } from '../../utils/apiUtils';
+
+export default (res: NextApiResponse): void => {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/api/capConfirmation.ts:4]: could we include req here also? but like this 
```suggestion
export default (_req: NextApiRequestWithSession, res: NextApiResponse): void => {
```

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+import { NextPageContextWithSession, ConfirmationElement, Cap } from '../interfaces';
+import TwoThirdsLayout from '../layout/Layout';
+import CsrfForm from '../components/CsrfForm';
+import ConfirmationTable from '../components/ConfirmationTable';
+import { getSessionAttribute } from '../utils/sessions';
+import { isCapExpiry, isCapStartInfo, isWithErrors } from '../interfaces/typeGuards';
+import { getCsrfToken, sentenceCaseString, getAndValidateNoc } from '../utils';
+import { getProductGroupByNocAndId } from '../data/auroradb';
+import { CapStartInfo } from 'src/interfaces/matchingJsonTypes';
+import { isServiceListAttributeWithErrors } from './serviceList';
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:21]:
```suggestion
import { isServiceListAttributeWithErrors } from '../../src/pages/serviceList';
```

File: repos/fdbt-site/tests/pages/api/typeOfCap.test.ts (hunk 4 above)
Comment (NitishAroraKPMG) [reply repos/fdbt-site/tests/pages/api/typeOfCap.test.ts:77]: it was written earlier, hence updated and will check the content then

File: repos/fdbt-site/src/pages/capConfirmation.tsx (hunk 6 above)
Comment (NitishAroraKPMG) [reply repos/fdbt-site/src/pages/capConfirmation.tsx:86]: these are hard coded, do we still need the function, i think, better to use when we have dynamic text

File: repos/fdbt-site/src/pages/capConfirmation.tsx (hunk 16 above)
Comment (RobCattonKPMG) [reply repos/fdbt-site/src/pages/capConfirmation.tsx:21]: Are we going to change the code then?

File: repos/fdbt-site/src/pages/capConfirmation.tsx (hunk 1 above)
Comment (RobCattonKPMG) [reply repos/fdbt-site/src/pages/capConfirmation.tsx:176]: @NitishAroraKPMG  are we not going to do this?

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+    const serviceListAttribute = getSessionAttribute(ctx.req, SERVICE_LIST_ATTRIBUTE);
+
+    const multiTapsPricingAttribute = getSessionAttribute(ctx.req, MULTI_TAPS_PRICING_ATTRIBUTE);
+
+    const capDistancePricingAttribute = getSessionAttribute(ctx.req, CAP_PRICING_PER_DISTANCE_ATTRIBUTE);
+
+    const tapsPricingContents: string[] = [];
+    if (multiTapsPricingAttribute && !isWithErrors(multiTapsPricingAttribute)) {
+        Object.entries(multiTapsPricingAttribute.tapDetails).forEach((tap) => {
+            tapsPricingContents.push(`Tap number: ${Number(tap[0]) + 1}, Price: ${tap[1]}`);
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:205]: price needs a £

File: repos/fdbt-site/src/pages/capConfirmation.tsx [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+        });
+    }
+
+    let capDistancePricingContents: string[] = [];
+    const distanceBands: string[] = [];
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/capConfirmation.tsx:210]: If I select 'rolling days' in /defineCapStart, it is not shown in the confirmation page.

Comment (RobCattonKPMG) [reply repos/fdbt-site/src/pages/capConfirmation.tsx:210]: Also, if the user creates a 'capped by taps' product, and clicks the back button whilst on the caps confirmation page, the site errors.

**Review by RobCattonKPMG — APPROVED** [review]

**Pull request merged**